cheetah/accelerator/space_charge_kick.py

```diff
@@ -84,146 +86,162 @@ def __init__(
-                & (idx_x < self.grid_shape[0])
-                & (idx_y >= 0)
-                & (idx_y < self.grid_shape[1])
-                & (idx_s >= 0)
-                & (idx_s < self.grid_shape[2])
-            )
+        charge = torch.zeros(
+            beam.particles.shape[:-2] + self.grid_shape, **self.factory_kwargs
+        )
 
-            # Accumulate the charge contributions
-            repeated_charges = particle_charge.repeat_interleave(
-                8
-            )  # Shape:(8*n_particles,)
-            values = (cell_weights.view(-1) * repeated_charges)[valid_mask]
-            charge[i_batch].index_put_(
-                (idx_x[valid_mask], idx_y[valid_mask], idx_s[valid_mask]),
-                values,
-                accumulate=True,
-            )
+        # Get particle positions
+        particle_positions = moments[..., [0, 2, 4]]
+        normalized_positions = (
+            particle_positions + grid_dimensions.unsqueeze(-2)
+        ) / cell_size.unsqueeze(-2)
 
-        # End of loop over batch
-        inv_cell_volume = 1 / (cell_size[:, 0] * cell_size[:, 1] * cell_size[:, 2])
+        # Find indices of the lower corners of the cells containing the particles
+        cell_indices = torch.floor(normalized_positions).type(torch.long)
+
+        # Calculate the weights for all surrounding cells
+        offsets = torch.tensor(
+            [
+                [0, 0, 0],
+                [0, 0, 1],
+                [0, 1, 0],
+                [0, 1, 1],
+                [1, 0, 0],
+                [1, 0, 1],
+                [1, 1, 0],
+                [1, 1, 1],
+            ]
+        )
+        surrounding_indices = cell_indices.unsqueeze(-2) + offsets.unsqueeze(-3)
+        # Shape: (..., num_particles, 8, 3)
+        weights = 1 - torch.abs(
+            normalized_positions.unsqueeze(-2) - surrounding_indices
+        )
+        # Shape: (.., num_particles, 8, 3)
+        cell_weights = weights.prod(dim=-1)  # Shape: (.., num_particles, 8)
+
+        # Add the charge contributions to the cells
+        idx_x = surrounding_indices[..., 0].flatten(start_dim=-2)
+        idx_y = surrounding_indices[..., 1].flatten(start_dim=-2)
+        idx_s = surrounding_indices[..., 2].flatten(start_dim=-2)
+        # Shape: (..., 8 * num_particles)
+
+        # Check that particles are inside the grid
+        valid_mask = (
+            (idx_x >= 0)
+            & (idx_x < self.grid_shape[0])
+            & (idx_y >= 0)
+            & (idx_y < self.grid_shape[1])
+            & (idx_s >= 0)
+            & (idx_s < self.grid_shape[2])
+        )
+
+        # Accumulate the charge contributions
+        repeated_charges = beam.particle_charges.repeat_interleave(
+            repeats=8, dim=-1
+        )  # Shape:(..., 8 * num_particles)
+        values = (cell_weights.flatten(start_dim=-2) * repeated_charges)[valid_mask]
+        charge[..., idx_x[valid_mask], idx_y[valid_mask], idx_s[valid_mask]] += values
```

Comment:
I tracked down the issue to this line, which is not producing the same result as the previous inplace `index_put_`

Reply:
I surprised right now that that's still in there ... I thought I reverted that 🤔

Reply:
@cr-xu do you happen to have caught how the values differed?

Reply:
So a simple `+=` doesn't handle values added to repeated indices, this line basically only adds charges once to a grid, causing the SC force to be way smaller thant it should be

Reply:
Just putting the original implementation here for reference:

```python
charge[i_batch].index_put_(
    (idx_x[valid_mask], idx_y[valid_mask], idx_s[valid_mask]),
    values,
    accumulate=True,
)
```

Reply:
I think [`scatter_add_`](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_) would be a replacement, which has the backpass capability

Reply:
I'm trying to understand the differences between [`scatter_add_`](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_) and [`index_put_`](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.index_put_). The problem with the original `index_put_` is that it doesn't yet work on the vectorised tensors and I'm not quite sure how it would be made to do that.

Reply:
I think the crux is that the vector indices need to be correctly included in the index. If the it done, we can probably just revert to the old `index_put_` line.